crates/bevy_math/src/primitives/dim3.rs

```diff
@@ -0,0 +1,172 @@
+    /// the positive and negative direction
+    pub half_length: f32,
+}
+impl Primitive3d for Segment3d {}
+
+impl Segment3d {
+    /// Create a line segment from a direction and full length of the line
+    pub fn new(direction: Direction3d, length: f32) -> Self {
+        Self {
+            direction,
+            half_length: length / 2.,
+        }
+    }
+
+    /// Get a line segment and translation from two points at each end of a line segment
+    ///
+    /// Panics if point1 == point2
+    pub fn from_points(point1: Vec3, point2: Vec3) -> (Self, Vec3) {
+        let diff = point2 - point1;
+        let length = diff.length();
+        (
+            Self::new(Direction3d::from_normalized(diff / length), length),
+            (point1 + point2) / 2.,
+        )
+    }
+
+    /// Get the position of the first point on the line segment
+    pub fn point1(&self) -> Vec3 {
+        *self.direction * -self.half_length
+    }
+
+    /// Get the position of the second point on the line segment
+    pub fn point2(&self) -> Vec3 {
+        *self.direction * self.half_length
+    }
+}
+
+/// A series of connected line segments in 3D space.
+///
+/// For a version without generics: [`BoxedPolyline3d`]
+#[derive(Clone, Debug)]
+pub struct Polyline3d<const N: usize> {
+    /// The vertices of the polyline
+    pub vertices: [Vec3; N],
+}
+impl<const N: usize> Primitive3d for Polyline3d<N> {}
+
+/// A series of connected line segments in 3D space.
+///
+/// For a version without alloc: [`Polyline3d`]
+#[derive(Clone, Debug)]
+pub struct BoxedPolyline3d {
+    /// The vertices of the polyline
+    pub vertices: Box<[Vec3]>,
+}
+impl Primitive3d for BoxedPolyline3d {}
+
+/// A cuboid primitive, more commonly known as a box.
+#[derive(Clone, Copy, Debug)]
+pub struct Cuboid {
+    /// Half of the width, height and depth of the cuboid
+    pub half_extents: Vec3,
+}
+impl Primitive3d for Cuboid {}
+
+impl Cuboid {
+    /// Create a cuboid from a full x, y and z length
+    pub fn new(x_length: f32, y_length: f32, z_length: f32) -> Self {
+        Self::from_size(Vec3::new(x_length, y_length, z_length))
+    }
+
+    /// Create a cuboid from the full size of the cuboid
+    pub fn from_size(size: Vec3) -> Self {
+        Self {
+            half_extents: size / 2.,
+        }
+    }
+}
+
+/// A cylinder primitive
+#[derive(Clone, Copy, Debug)]
+pub struct Cylinder {
+    /// The radius of the cylinder
+    pub radius: f32,
+    /// The half height of the cylinder
+    pub half_height: f32,
+}
+impl Primitive3d for Cylinder {}
+
+impl Cylinder {
+    /// Create a cylinder from a radius and full height
+    pub fn new(radius: f32, height: f32) -> Self {
+        Self {
+            radius,
+            half_height: height / 2.,
+        }
+    }
+}
+
+/// A capsule primitive.
+/// A capsule is defined as a surface at a distance (radius) from a line
+#[derive(Clone, Copy, Debug)]
+pub struct Capsule {
```

Comment:
This probably isn't really an issue, but using the same capsule shape for 2D and 3D might make some helper methods a bit more annoying. For example, `capsule.area()` is different for 2D and 3D since the surface area is interpreted differently, so there'd need to be separate `area_2d` and `area_3d` methods.